### ISSUE TITLE
Testsuite for Solaris when building 32 bit breaks

### ISSUE THREAD

**Issue by dago** [collectd/collectd#1301]
When building the latest version on Solaris 10 with 32 bit the test suite does not compile:
```
gmake  test_common test_meta_data test_utils_avltree test_utils_heap test_utils_time test_utils_subst
gmake[4]: Entering directory '/home/dam/work/collectd/src/daemon'
  CC       common_test.o
In file included from collectd.h:31:0,
                 from common.h:31,
                 from common_test.c:28:
../../src/config.h:1622:0: error: "_FILE_OFFSET_BITS" redefined [-Werror]
 #define _FILE_OFFSET_BITS 64
 ^
In file included from /usr/include/inttypes.h:20:0,
                 from ../../src/testing.h:27,
                 from common_test.c:27:
/opt/csw/lib/gcc/i386-pc-solaris2.10/4.9.2/include-fixed/sys/feature_tests.h:196:0: note: this is the location of the previous definition
 #define _FILE_OFFSET_BITS 32
 ^
cc1: all warnings being treated as errors
Makefile:986: recipe for target 'common_test.o' failed
```

This has probably to do with largefiles as described in lf64(5) being different in the standard GCC headers and the determined width.

**Comment by mfournier** [comment]
@dago are you able to retest current master by any chance ? It looks like #1442 should have fixed this. Thanks !

**Comment by dago** [comment]
Sure, this is integrated in our buildbot:
  https://buildfarm.opencsw.org/buildbot/waterfall?category=collectd
If you go back in history you see that is passes on the 57dfc4722bb1c4d7f735c71aa113b9f6bbdce964  commit:
  https://buildfarm.opencsw.org/buildbot/builders/collectd-solaris10-i386
That means the issue is fixed now, but not with the commit from #1442.

Feel free to add a link to our buildbot for CollectD to your development page:
  https://collectd.org/dev-info.shtml

Best regards -- Dago

**Comment by mfournier** [comment]
Great, thanks @dago !

**Comment by dago** [comment]
The issue reappeared in the latest master:
```
  CCLD     write_http.la
gmake  test_format_json test_utils_latency test_utils_vl_lookup test_utils_mount test_plugin_ceph
gmake[4]: Entering directory '/home/dam/work/collectd/src'
  CC       utils_format_json_test.o
In file included from ./daemon/collectd.h:31:0,
                 from utils_format_json_test.c:28:
./config.h:1671:0: error: "_FILE_OFFSET_BITS" redefined [-Werror]
 #define _FILE_OFFSET_BITS 64
 ^
In file included from /usr/include/inttypes.h:20:0,
                 from testing.h:30,
                 from utils_format_json_test.c:27:
/opt/csw/lib/gcc/sparc-sun-solaris2.10/5.2.0/include-fixed/sys/feature_tests.h:196:0: note: this is the location of the previous definition
 #define _FILE_OFFSET_BITS 32
 ^
utils_format_json_test.c:45:3: error: conflicting types for 'label_t'
 } label_t;
   ^
In file included from /usr/include/sys/types.h:23:0,
                 from ./daemon/collectd.h:36,
                 from utils_format_json_test.c:28:
/usr/include/sys/machtypes.h:33:42: note: previous declaration of 'label_t' was here
 typedef struct _label_t { long val[2]; } label_t;
                                          ^
cc1: all warnings being treated as errors
Makefile:4130: recipe for target 'utils_format_json_test.o' failed
gmake[4]: *** [utils_format_json_test.o] Error 1
```

**Comment by rubenk** [comment]
Thanks for letting us know @dago!

While we could shuffle some headers around again, I'd like to fix this once and for all and get to the bottom of it.

I'm curious as to why your environment sets `#define _FILE_OFFSET_BITS 32` and not 64.
Does your environment not support large files? Or perhaps it needs a preprocessor define to expose support for it.

Could you paste a few lines around /opt/csw/lib/gcc/sparc-sun-solaris2.10/5.2.0/include-fixed/sys/feature_tests.h:196 here?

**Comment by rubenk** [comment]
Never mind, the problem is that we set _FILE_OFFSET_BITS in `config.h`, which is included by `collectd.h`, but not by `testing.h`.` testing.h` includes `<inttypes.h>`, which includes `feature_tests.h`, which defines _FILE_OFFSET_BITS to 32.

I'll add `config.h` to `testing.h`, that should fix it.

**Comment by rubenk** [comment]
The other issue is that the system headers also define a label_t type. Why they do that is a mystery to me, they should have prefixed it with an underscore..

**Comment by dago** [comment]
The corresponding code from `/usr/include/sys/machtypes.h` looks like this:
```
#if (!defined(_POSIX_C_SOURCE) && !defined(_XOPEN_SOURCE)) || \
        defined(__EXTENSIONS__)

typedef struct  _label_t { long val[2]; } label_t;

#endif /* !defined(_POSIX_C_SOURCE)... */
```

**Comment by rubenk** [comment]
Ok, I think we should set _POSIX_C_SOURCE. We now only set _POSIX_THREAD_SEMANTICS (in configure.ac), but https://docs.oracle.com/cd/E19683-01/806-6867/6jfpgdcob/index.html#compile-27862 suggest the former is better.

**Comment by rubenk** [comment]
Hmm this breaks python, because it knows better:

```
In file included from ./daemon/collectd.h:31:0,
                 from python.c:32:
./config.h:1683:0: error: "_REENTRANT" redefined [-Werror]
 #define _REENTRANT /**/
 ^
In file included from /usr/include/python2.6/Python.h:8:0,
                 from python.c:27:
/usr/include/python2.6/pyconfig.h:999:0: note: this is the location of the previous definition
 #define _REENTRANT 1
 ^
```
https://bugs.python.org/issue1045893

Is it too early in the morning to start drinking yet?

**Comment by rubenk** [comment]
Ok, I got as far as setting _POSIX_C_SOURCE. This unfortunately sets _XPG4_2 which hides IPV6_MULTICAST_HOPS. I can fix that by adding __EXTENSIONS__, but that in turn causes label_t to be defined.

So it seems to me there's no really good solution here. @octo any objections to renaming label_t to something else?

**Comment by octo** [comment]
From my limited context, that sounds like the most reasonable solution.

**Comment by octo** [comment]
> The other issue is that the system headers also define a label_t type. Why they do that is a mystery to me, they should have prefixed it with an underscore..
+1, why have a reserved namespace if you're not using it?

**Comment by rubenk** [comment]
At least it seems they fixed that mistake in Solaris 11.

**Comment by dago** [comment]
Would be nice to apply this at least to the 5.6 branch and master.

**Comment by rubenk** [comment]
@dago it already is, or am I missing something?

**Comment by dago** [comment]
Nope, test suite still fails, see latest log from the test suites:
  https://buildfarm.opencsw.org/buildbot/waterfall?category=collectd

**Comment by rubenk** [comment]
Both branches have that patch:
```
$ git branch --contains f7e7130af45d2320d5bf1a4c7cf6f553f213f1fc
* collectd-5.6
  master
```
 Seems it doesn't work though, I'd swear I tested this.

**Comment by rubenk** [comment]
@dago should be fixed by 7a43e0b7b46589add8b7a3e7a870dc6485e04cdf, for real this time. Thanks for noticing!

**Comment by dago** [comment]
Yes, everything looks good now. Thanks!
